File: packages/agent-runtime/py/src/steerable_agent_runtime/skills.py
"""Skill seam — layered disclosure of SKILL.md prompt modules.

A *skill* is a directory containing a ``SKILL.md``: YAML-ish frontmatter
(name, description, priority, conditions, …) plus a markdown body of
instructions. Products inject foundational skills into the system prompt
directly; everything else is disclosed progressively — the model sees a
one-line-per-skill catalog and calls the ``skill`` tool to load a body on
demand. This is the pattern codex / Claude Code / Cursor converge on, and
it is what keeps a large skill library from blowing the prompt budget.

Three roles (capability seam):

- **Definition** — ``SkillSummary`` / ``SkillDefinition`` and the
  ``SkillProvider`` protocol.
- **Provider** — ``FilesystemSkillProvider`` reads ``<root>/<dir>/SKILL.md``.
  Frontmatter parsing is field-compatible with deeppath-agent's
  ``skill-loader.ts`` (same minimal YAML subset, same name validation,
  same user-root-overrides-builtin semantics) and adds two fields:
  ``layer: eager|catalog`` (default derived from priority) and the
  ecosystem's ``disable-model-invocation: true`` (Claude/codex) mapped to
  ``model_invocable=False``.
- **Consumer** — ``skill_tool_descriptor()`` (OpenAI schema to append to
  the tools list), ``SkillExecutor`` (``ToolExecutor`` decorator answering
  ``skill`` calls with the body wrapped in ``<skill_content>`` markers),
  and ``SkillHooks`` (``pre_step`` injects the catalog into the system
  message on the first round — a transcript rewrite, so the injection is
  recorded as a ``hook_action`` event: model-visible ⟺ logged).

Opt-in: hosts that don't want skills pass no provider and nothing changes.
"""

from __future__ import annotations

import logging
import re
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Literal, Protocol, Sequence, runtime_checkable

from steerable_agent_protocol.generated import ToolCall, ToolResult

from .hooks import NoopHooks, PreStepAction
from .llm import LLMMessage
from .loop import LoopContext, ToolExecutor

logger = logging.getLogger(__name__)

#: Priority at or above which a skill defaults to the eager layer (always
#: injected into the system prompt). Chosen so deeppath-agent's existing
#: built-ins keep their current behavior without frontmatter changes:
#: identity 1000 / plan-mode 950 / tool-usage 900 / anti-deferred 880 /
#: data-grounding 875 stay eager; proactive-coding 705 / local-exec 700 /
#: cflog 600 and user imports (default 500) become catalog.
EAGER_PRIORITY_THRESHOLD = 850

_NAME_RE = re.compile(r"^[a-z0-9]([a-z0-9-]*[a-z0-9])?$")
_NAME_MAX_LEN = 64
_DESC_MAX_LEN = 1024

Layer = Literal["eager", "catalog"]


# ---------------------------------------------------------------------------
# Definition
# ---------------------------------------------------------------------------


@dataclass(frozen=True, slots=True)
class SkillSummary:
    """Everything about a skill except its body (cheap to list)."""

    name: str
    description: str
    display_name: str
    priority: int
    layer: Layer
    #: False = user-invocable only (``/name``); hidden from the catalog and
    #: rejected by ``SkillExecutor`` (ecosystem ``disable-model-invocation``).
    model_invocable: bool
    conditions: tuple[str, ...]
    match: Literal["any", "all"]
    dir_name: str


@dataclass(frozen=True, slots=True)
class SkillDefinition(SkillSummary):
    """A skill with its body; ``content`` has ``{scripts}`` resolved."""

    content: str
    root: str


@runtime_checkable
class SkillProvider(Protocol):
    """Source of skills. ``list`` is the catalog view; ``get`` loads a body.

    ``get`` accepts the skill's ``name``, directory name, or display name
    (case-insensitive) — the same aliases the desktop's ``/`` trigger and
    forced-skill path accept, so a name the model copies out of the catalog
    always resolves.
    """

    def list(self) -> Sequence[SkillSummary]: ...

    def get(self, name: str) -> SkillDefinition | None: ...


def matches_conditions(skill: SkillSummary, active: set[str]) -> bool:
    """Condition gate, ported from the TS loader: no conditions = always;
    ``match: all`` requires every condition, otherwise any one suffices."""
    if not skill.conditions:
        return True
    if skill.match == "all":
        return all(c in active for c in skill.conditions)
    return any(c in active for c in skill.conditions)


def select_catalog(
    skills: Sequence[SkillSummary],
    conditions: set[str] | frozenset[str] = frozenset(),
    exclude: Sequence[str] = (),
    ignore_conditions: bool = False,
) -> list[SkillSummary]:
    """The model-visible catalog: catalog-layer, condition-matching,
    model-invocable skills, minus the host's exclusions (e.g. plan mode
    drops execution-oriented skills). ``ignore_conditions`` lists every
    catalog skill regardless of conditions (the desktop's all-round
    assistant persona opts into everything)."""
    active = set(conditions)
    excluded = {e.lower().strip() for e in exclude}

    def is_excluded(s: SkillSummary) -> bool:
        return (
            s.name.lower() in excluded
            or s.dir_name.lower() in excluded
            or (s.display_name != "" and s.display_name.lower() in excluded)
        )

    return [
        s
        for s in skills
        if s.layer == "catalog"
        and s.model_invocable
        and not is_excluded(s)
        and (ignore_conditions or matches_conditions(s, active))
    ]


def render_skill_catalog(
    skills: Sequence[SkillSummary], *, tool_name: str = "skill"
) -> str:
    """System-prompt section listing the catalog layer, one line per skill."""
    lines = [
        "# Available skills (load on demand)",
        "",
        "The skills below may be relevant to this turn; their full instructions "
        f"are not loaded yet. When the task matches one, call the `{tool_name}` "
        "tool with its `name` to load the instructions, then follow them. If "
        "none match, ignore this list — do not guess unlisted skill names.",
        "",
    ]
    for s in skills:
        label = s.name if not s.display_name else f"{s.name}({s.display_name})"
        lines.append(f"- {label}: {s.description}" if s.description else f"- {label}")
    return "\n".join(lines)


# ---------------------------------------------------------------------------
# Provider: filesystem
# ---------------------------------------------------------------------------


class FilesystemSkillProvider:
    """Reads ``<root>/<dir>/SKILL.md`` from each root, in order.

    Later roots override earlier ones on the same (case-insensitive) name —
    pass the built-in root first and the user root second to mirror the
    desktop's user-overrides-builtin semantics. Discovery happens once at
    construction; build a fresh provider per turn (skills can be imported
    mid-session).
    """

    def __init__(self, roots: Sequence[str | Path]) -> None:
        merged: dict[str, SkillDefinition] = {}
        for root in roots:
            for definition in _load_root(Path(root)):
                merged[definition.name.lower()] = definition
        self._skills = sorted(merged.values(), key=lambda d: d.dir_name)

    def list(self) -> Sequence[SkillSummary]:
        return self._skills

    def get(self, name: str) -> SkillDefinition | None:
        key = name.lower().strip()
        for skill in self._skills:
            if (
                skill.name.lower() == key
                or skill.dir_name.lower() == key
                or (skill.display_name != "" and skill.display_name.lower() == key)
            ):
                return skill
        return None


def _load_root(root: Path) -> list[SkillDefinition]:
    try:
        entries = sorted(p for p in root.iterdir() if p.is_dir())
    except OSError:
        logger.warning("skill root missing: %s", root)
        return []
    out: list[SkillDefinition] = []
    for skill_dir in entries:
        definition = _parse_skill_dir(skill_dir)
        if definition is not None:
            out.append(definition)
    return out


def _parse_skill_dir(skill_dir: Path) -> SkillDefinition | None:
    skill_file = skill_dir / "SKILL.md"
    try:
        if not skill_file.is_file():
            return None
        raw = skill_file.read_text(encoding="utf-8")
    except OSError:
        return None

    fm_raw = ""
    body = raw
    if raw.startswith("---"):
        rest = raw[3:]
        end = rest.find("\n---")
        if end != -1:
            fm_raw = rest[:end]
            body = re.sub(r"^\r?\n", "", rest[end + 4 :])

    content = body.strip()
    if not content:
        return None

    fm = _parse_frontmatter(fm_raw) if fm_raw.strip() else {}

    name = fm.get("name") if isinstance(fm.get("name"), str) else ""
    if not name:
        logger.warning("skill missing name: dir=%s", skill_dir.name)
        return None
    if not _valid_name(name, str(skill_file)):
        return None

    description = fm.get("description") if isinstance(fm.get("description"), str) else ""
    if len(description) > _DESC_MAX_LEN:
        logger.warning(
            "skill description too long: name=%s len=%d max=%d",
            name,
            len(description),
            _DESC_MAX_LEN,
        )

    priority_raw = fm.get("priority")
    priority = int(priority_raw) if isinstance(priority_raw, (int, float)) else 500

    conditions_raw = fm.get("conditions")
    conditions = tuple(conditions_raw) if isinstance(conditions_raw, list) else ()
    match: Literal["any", "all"] = (
        "all" if str(fm.get("match") or "").lower() == "all" else "any"
    )

    layer_raw = str(fm.get("layer") or "").lower()
    layer: Layer
    if layer_raw in ("eager", "catalog"):
        layer = layer_raw  # type: ignore[assignment]
    else:
        layer = "eager" if priority >= EAGER_PRIORITY_THRESHOLD else "catalog"

    # Ecosystem compat: Claude Code / codex mark user-only skills with
    # `disable-model-invocation: true`; they stay `/`-triggerable but leave
    # the catalog and get rejected by the skill tool.
    model_invocable = fm.get("disable-model-invocation") is not True

    # Same `{scripts}` resolution as the TS prompt assembly: point at the
    # skill's own scripts/ directory with uniform separators.
    scripts_path = str(skill_dir / "scripts")
    content = re.sub(r"\{scripts\}[/\\]", scripts_path + "/", content)
    content = content.replace("{scripts}", scripts_path)

    return SkillDefinition(
        name=name,
        description=description,
        display_name=(fm.get("displayName") or "").strip()
        if isinstance(fm.get("displayName"), str)
        else "",
        priority=priority,
        layer=layer,
        model_invocable=model_invocable,
        conditions=conditions,
        match=match,
        content=content,
        dir_name=skill_dir.name,
        root=str(skill_dir.parent),
    )


def _valid_name(name: str, source: str) -> bool:
    if not name or len(name) > _NAME_MAX_LEN:
        logger.warning("skill_name_invalid length=%d source=%s", len(name), source)
        return False
    if "--" in name:
        logger.warning("skill_name_consecutive_hyphens name=%s source=%s", name, source)
        return False
    if not _NAME_RE.match(name):
        logger.warning("skill_name_bad_format name=%s source=%s", name, source)
        return False
    return True


def _parse_frontmatter(raw: str) -> dict[str, Any]:
    """Minimal YAML-subset frontmatter parser — a 1:1 port of the TS loader's
    (flat keys only: scalars, quoted strings, inline ``[a, b]`` arrays,
    booleans, numbers; comments and blank lines skipped)."""
    out: dict[str, Any] = {}
    for raw_line in raw.split("\n"):
        trimmed = raw_line.rstrip("\r").strip()
        if not trimmed or trimmed.startswith("#"):
            continue
        colon = trimmed.find(":")
        if colon == -1:
            continue
        key = trimmed[:colon].strip()
        value = trimmed[colon + 1 :].strip()
        if not value:
            continue
        out[key] = _parse_value(value)
    return out


def _parse_value(value: str) -> Any:
    if value.startswith("[") and value.endswith("]"):
        inner = value[1:-1].strip()
        if not inner:
            return []
        return [s for s in (_unquote(p.strip()) for p in inner.split(",")) if s]
    if (value.startswith('"') and value.endswith('"')) or (
        value.startswith("'") and value.endswith("'")
    ):
        return _unquote(value)
    if value in ("true", "false"):
        return value == "true"
    if re.fullmatch(r"-?\d+(\.\d+)?", value):
        return float(value) if "." in value else int(value)
    return value


def _unquote(s: str) -> str:
    if len(s) >= 2 and s[0] == s[-1] and s[0] in ("'", '"'):
        return s[1:-1]
    return s


# ---------------------------------------------------------------------------
# Consumer: tool descriptor + executor decorator + catalog hook
# ---------------------------------------------------------------------------


@dataclass(frozen=True, slots=True)
class SkillConfig:
    """Tunables for the skill tool exposed by ``SkillExecutor``."""

    tool_name: str = "skill"
    description: str = (
        "Load a skill's full instructions by name. Call this when the task "
        "matches a skill from the available-skills list in the system prompt, "
        "then follow the loaded instructions."
    )


def skill_tool_descriptor(config: SkillConfig | None = None) -> dict[str, Any]:
    """OpenAI tool schema to append to the loop's tools list."""
    config = config or SkillConfig()
    return {
        "type": "function",
        "function": {
            "name": config.tool_name,
            "description": config.description,
            "parameters": {
                "type": "object",
                "properties": {
                    "name": {
                        "type": "string",
                        "description": (
                            "Skill name exactly as listed in the available-skills "
                            "catalog."
                        ),
                    },
                },
                "required": ["name"],
                "additionalProperties": False,
            },
        },
    }


class SkillExecutor:
    """ToolExecutor decorator: ``config.tool_name`` calls load a skill body.

    The body is returned wrapped in ``<skill_content name="...">`` markers
    (dsh's convention) so both the model and trace readers can tell loaded
    instructions apart from tool data.
    """

    def __init__(
        self,
        inner: ToolExecutor,
        provider: SkillProvider,
        config: SkillConfig | None = None,
        *,
        conditions: set[str] | frozenset[str] = frozenset(),
        exclude: Sequence[str] = (),
        ignore_conditions: bool = False,
    ) -> None:
        self._inner = inner
        self._provider = provider
        self._config = config or SkillConfig()
        self._conditions = conditions
        self._exclude = tuple(exclude)
        self._ignore_conditions = ignore_conditions

    async def execute(self, call: ToolCall, ctx: LoopContext) -> ToolResult:
        if call.name != self._config.tool_name:
            return await self._inner.execute(call, ctx)
        name = str((call.arguments or {}).get("name") or "").strip()
        if not name:
            return ToolResult(
                success=False,
                error="missing required argument: name",
                needsFollowup=True,
            )
        skill = self._provider.get(name)
        if skill is None:
            available = [
                s.name
                for s in select_catalog(
                    self._provider.list(),
                    self._conditions,
                    self._exclude,
                    self._ignore_conditions,
                )
            ]
            error = f"Unknown skill: {name}"
            if available:
                error += f". Available skills: {', '.join(available)}"
            return ToolResult(
                success=False,
                error=error,
                needsFollowup=True,
                data={"unknownSkill": name},
            )
        excluded = {e.lower().strip() for e in self._exclude}
        if (
            skill.name.lower() in excluded
            or skill.dir_name.lower() in excluded
            or (skill.display_name != "" and skill.display_name.lower() in excluded)
        ):
            return ToolResult(
                success=False,
                error=f"Skill '{skill.name}' is not available in this mode.",
                needsFollowup=True,
                data={"skill": skill.name, "excluded": True},
            )
        if not skill.model_invocable:
            return ToolResult(
                success=False,
                error=(
                    f"Skill '{skill.name}' is user-invocable only "
                    "(disable-model-invocation); do not load it yourself."
                ),
                needsFollowup=True,
                data={"skill": skill.name, "modelInvocable": False},
            )
        return ToolResult(
            success=True,
            message=(
                f'<skill_content name="{skill.name}">\n{skill.content}\n</skill_content>'
            ),
            data={"skill": skill.name},
        )

    def concurrency_safe(self, call: ToolCall) -> bool:
        # Loading a skill is a pure read — safe to batch with siblings;
        # other calls defer to the inner executor's own judgement.
        if call.name == self._config.tool_name:
            return True
        inner_safe = getattr(self._inner, "concurrency_safe", None)
        return bool(inner_safe and inner_safe(call))


class SkillHooks(NoopHooks):
    """LoopHooks: inject the skill catalog into the system message, once.

    The injection is a first-round ``pre_step`` transcript rewrite, so the
    loop records it as a ``hook_action`` event (``action: skill_catalog``)
    and the catalog is reconstructable from the trace. Compaction preserves
    system messages, so the catalog survives later rewrites.
    """

    def __init__(
        self,
        provider: SkillProvider,
        *,
        conditions: set[str] | frozenset[str] = frozenset(),
        exclude: Sequence[str] = (),
        config: SkillConfig | None = None,
        ignore_conditions: bool = False,
    ) -> None:
        self._provider = provider
        self._conditions = conditions
        self._exclude = tuple(exclude)
        self._config = config or SkillConfig()
        self._ignore_conditions = ignore_conditions
        self._injected = False

    async def pre_step(
        self, transcript: list[LLMMessage], ctx: LoopContext
    ) -> PreStepAction:
        if self._injected:
            return PreStepAction(kind="proceed")
        self._injected = True
        catalog = select_catalog(
            self._provider.list(), self._conditions, self._exclude, self._ignore_conditions
        )
        if not catalog:
            return PreStepAction(kind="proceed")
        section = render_skill_catalog(catalog, tool_name=self._config.tool_name)
        rewritten = list(transcript)
        if rewritten and rewritten[0].role == "system":
            first = rewritten[0]
            rewritten[0] = LLMMessage(
                role="system",
                content=f"{first.content}\n\n{section}",
                name=first.name,
            )
        else:
            rewritten.insert(0, LLMMessage(role="system", content=section))
        return PreStepAction(
            kind="proceed",
            transcript=rewritten,
            reason=f"skill catalog injected ({len(catalog)} skills)",
            rewrite_action="skill_catalog",
        )

File: packages/agent-runtime/rs/src/skills.rs
//! SKILL.md listing for `skills.list` (Python `steerable_agent_runtime.skills`).

use std::fs;
use std::path::{Path, PathBuf};

use serde_json::{json, Map, Value};

pub const EAGER_PRIORITY_THRESHOLD: i64 = 850;
const NAME_MAX_LEN: usize = 64;
const DESC_MAX_LEN: usize = 1024;
pub const DEFAULT_MAX_CATALOG_SKILLS: usize = 50;

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct SkillDefinition {
    pub name: String,
    pub description: String,
    pub display_name: String,
    pub priority: i64,
    pub layer: String,
    pub model_invocable: bool,
    pub conditions: Vec<String>,
    pub match_mode: String,
    pub dir_name: String,
    pub content: String,
    pub root: String,
    pub tags: Vec<String>,
}

impl SkillDefinition {
    pub fn to_rpc(&self) -> Value {
        json!({
            "name": self.name,
            "displayName": self.display_name,
            "description": self.description,
            "priority": self.priority,
            "tags": self.tags,
            "conditions": self.conditions,
            "match": self.match_mode,
            "layer": self.layer,
            "modelInvocable": self.model_invocable,
            "content": self.content,
            "dirName": self.dir_name,
            "skillsDir": self.root,
        })
    }
}

pub fn matches_conditions(skill: &SkillDefinition, active: &[String]) -> bool {
    if skill.conditions.is_empty() {
        return true;
    }
    if skill.match_mode == "all" {
        skill
            .conditions
            .iter()
            .all(|condition| active.iter().any(|item| item == condition))
    } else {
        skill
            .conditions
            .iter()
            .any(|condition| active.iter().any(|item| item == condition))
    }
}

fn is_excluded(skill: &SkillDefinition, excluded: &[String]) -> bool {
    let names = [
        skill.name.to_ascii_lowercase(),
        skill.dir_name.to_ascii_lowercase(),
        skill.display_name.to_ascii_lowercase(),
    ];
    excluded
        .iter()
        .any(|item| names.iter().any(|name| !name.is_empty() && name == item))
}

pub fn select_skills(
    skills: &[SkillDefinition],
    conditions: &[String],
    exclude: &[String],
    ignore_conditions: bool,
) -> Vec<SkillDefinition> {
    let excluded: Vec<String> = exclude
        .iter()
        .map(|item| item.trim().to_ascii_lowercase())
        .collect();
    skills
        .iter()
        .filter(|skill| {
            !is_excluded(skill, &excluded)
                && (ignore_conditions || matches_conditions(skill, conditions))
        })
        .cloned()
        .collect()
}

pub fn select_catalog(
    skills: &[SkillDefinition],
    conditions: &[String],
    exclude: &[String],
    ignore_conditions: bool,
) -> Vec<SkillDefinition> {
    select_skills(skills, conditions, exclude, ignore_conditions)
        .into_iter()
        .filter(|skill| skill.layer == "catalog" && skill.model_invocable)
        .collect()
}

pub fn render_skill_catalog(
    skills: &[SkillDefinition],
    tool_name: &str,
    max_skills: usize,
) -> String {
    let mut lines = vec![
        "# Available skills (load on demand)".to_string(),
        String::new(),
        format!(
            "The skills below may be relevant to this turn; their full instructions are not loaded yet. When the task matches one, call the `{tool_name}` tool with its `name` to load the instructions, then follow them. If none match, ignore this list — do not guess unlisted skill names."
        ),
        String::new(),
    ];
    let mut ranked = skills.to_vec();
    ranked.sort_by_key(|skill| std::cmp::Reverse(skill.priority));
    let shown = ranked.iter().take(max_skills).collect::<Vec<_>>();
    for skill in &shown {
        let label = if skill.display_name.is_empty() {
            skill.name.clone()
        } else {
            format!("{}({})", skill.name, skill.display_name)
        };
        if skill.description.is_empty() {
            lines.push(format!("- {label}"));
        } else {
            lines.push(format!("- {label}: {}", skill.description));
        }
    }
    let omitted = ranked.len().saturating_sub(shown.len());
    if omitted > 0 {
        lines.push(format!(
            "- …and {omitted} more skills not listed (catalog capped at {max_skills}); ask if none of these fit."
        ));
    }
    lines.join("\n")
}

pub struct FilesystemSkillProvider {
    skills: Vec<SkillDefinition>,
}

impl FilesystemSkillProvider {
    pub fn new(roots: &[PathBuf]) -> Self {
        let mut merged: std::collections::HashMap<String, SkillDefinition> =
            std::collections::HashMap::new();
        for root in roots {
            for definition in load_root(root) {
                merged.insert(definition.name.to_ascii_lowercase(), definition);
            }
        }
        let mut skills: Vec<SkillDefinition> = merged.into_values().collect();
        skills.sort_by(|a, b| a.dir_name.cmp(&b.dir_name));
        Self { skills }
    }

    pub fn list(&self) -> &[SkillDefinition] {
        &self.skills
    }

    pub fn get(&self, name: &str) -> Option<&SkillDefinition> {
        let key = name.trim().to_ascii_lowercase();
        self.skills.iter().find(|skill| {
            skill.name.to_ascii_lowercase() == key
                || skill.dir_name.to_ascii_lowercase() == key
                || (!skill.display_name.is_empty()
                    && skill.display_name.to_ascii_lowercase() == key)
        })
    }
}

fn load_root(root: &Path) -> Vec<SkillDefinition> {
    let Ok(entries) = fs::read_dir(root) else {
        return Vec::new();
    };
    let mut dirs: Vec<PathBuf> = entries
        .filter_map(Result::ok)
        .map(|entry| entry.path())
        .filter(|path| path.is_dir())
        .collect();
    dirs.sort();
    dirs.into_iter()
        .filter_map(|path| parse_skill_dir(&path))
        .collect()
}

fn parse_skill_dir(skill_dir: &Path) -> Option<SkillDefinition> {
    let skill_file = skill_dir.join("SKILL.md");
    if !skill_file.is_file() {
        return None;
    }
    let raw = fs::read_to_string(&skill_file).ok()?;
    let (fm_raw, body) = split_frontmatter(&raw);
    let content = body.trim();
    if content.is_empty() {
        return None;
    }
    let fm = if fm_raw.trim().is_empty() {
        Map::new()
    } else {
        parse_frontmatter(&fm_raw)
    };
    let name = fm.get("name").and_then(Value::as_str).unwrap_or("");
    if name.is_empty() || !valid_name(name) {
        return None;
    }
    let description = fm
        .get("description")
        .and_then(Value::as_str)
        .unwrap_or("")
        .to_string();
    let _ = description.len() > DESC_MAX_LEN;
    let priority = match fm.get("priority") {
        Some(Value::Number(number)) => number
            .as_i64()
            .or_else(|| number.as_f64().map(|n| n as i64))
            .unwrap_or(500),
        _ => 500,
    };
    let conditions = string_list(fm.get("conditions"));
    let tags = string_list(fm.get("tags"));
    let match_mode = if fm
        .get("match")
        .and_then(Value::as_str)
        .unwrap_or("")
        .eq_ignore_ascii_case("all")
    {
        "all"
    } else {
        "any"
    }
    .to_string();
    let layer_raw = fm
        .get("layer")
        .and_then(Value::as_str)
        .unwrap_or("")
        .to_ascii_lowercase();
    let layer = if layer_raw == "eager" || layer_raw == "catalog" {
        layer_raw
    } else if priority >= EAGER_PRIORITY_THRESHOLD {
        "eager".into()
    } else {
        "catalog".into()
    };
    let model_invocable = fm.get("disable-model-invocation") != Some(&json!(true));
    let scripts_path = skill_dir
        .join("scripts")
        .to_string_lossy()
        .replace('\\', "/");
    let mut content = content.to_string();
    content = replace_scripts(&content, &scripts_path);
    let display_name = fm
        .get("displayName")
        .and_then(Value::as_str)
        .unwrap_or("")
        .trim()
        .to_string();
    Some(SkillDefinition {
        name: name.to_string(),
        description,
        display_name,
        priority,
        layer,
        model_invocable,
        conditions,
        match_mode,
        dir_name: skill_dir
            .file_name()
            .map(|name| name.to_string_lossy().into_owned())
            .unwrap_or_default(),
        content,
        root: skill_dir
            .parent()
            .map(|path| path.to_string_lossy().into_owned())
            .unwrap_or_default(),
        tags,
    })
}

fn split_frontmatter(raw: &str) -> (String, String) {
    if !raw.starts_with("---") {
        return (String::new(), raw.to_string());
    }
    let rest = &raw[3..];
    let Some(end) = rest.find("\n---") else {
        return (String::new(), raw.to_string());
    };
    let fm_raw = rest[..end].to_string();
    let mut body = rest[end + 4..].to_string();
    if body.starts_with('\r') {
        body = body[1..].to_string();
    }
    if body.starts_with('\n') {
        body = body[1..].to_string();
    }
    (fm_raw, body)
}

fn replace_scripts(content: &str, scripts_path: &str) -> String {
    let mut out = String::new();
    let bytes = content.as_bytes();
    let mut index = 0;
    let needle = b"{scripts}";
    while index < bytes.len() {
        if let Some(rel) = content[index..].find("{scripts}") {
            out.push_str(&content[index..index + rel]);
            let after = index + rel + needle.len();
            if content
                .as_bytes()
                .get(after)
                .is_some_and(|ch| *ch == b'/' || *ch == b'\\')
            {
                out.push_str(scripts_path);
                out.push('/');
                index = after + 1;
            } else {
                out.push_str(scripts_path);
                index = after;
            }
        } else {
            out.push_str(&content[index..]);
            break;
        }
    }
    out
}

fn valid_name(name: &str) -> bool {
    if name.is_empty() || name.len() > NAME_MAX_LEN || name.contains("--") {
        return false;
    }
    let bytes = name.as_bytes();
    let first = bytes[0];
    if !first.is_ascii_lowercase() && !first.is_ascii_digit() {
        return false;
    }
    if bytes.len() == 1 {
        return true;
    }
    let last = bytes[bytes.len() - 1];
    if !last.is_ascii_lowercase() && !last.is_ascii_digit() {
        return false;
    }
    bytes[1..bytes.len() - 1]
        .iter()
        .all(|ch| ch.is_ascii_lowercase() || ch.is_ascii_digit() || *ch == b'-')
}

fn parse_frontmatter(raw: &str) -> Map<String, Value> {
    let mut out = Map::new();
    for raw_line in raw.split('\n') {
        let trimmed = raw_line.trim_end_matches('\r').trim();
        if trimmed.is_empty() || trimmed.starts_with('#') {
            continue;
        }
        let Some(colon) = trimmed.find(':') else {
            continue;
        };
        let key = trimmed[..colon].trim();
        let value = trimmed[colon + 1..].trim();
        if value.is_empty() {
            continue;
        }
        out.insert(key.to_string(), parse_value(value));
    }
    out
}

fn parse_value(value: &str) -> Value {
    if value.starts_with('[') && value.ends_with(']') {
        let inner = value[1..value.len() - 1].trim();
        if inner.is_empty() {
            return json!([]);
        }
        let items: Vec<Value> = inner
            .split(',')
            .map(str::trim)
            .map(unquote)
            .filter(|item| !item.is_empty())
            .map(Value::from)
            .collect();
        return Value::Array(items);
    }
    if (value.starts_with('"') && value.ends_with('"'))
        || (value.starts_with('\'') && value.ends_with('\''))
    {
        return Value::from(unquote(value));
    }
    if value == "true" {
        return json!(true);
    }
    if value == "false" {
        return json!(false);
    }
    if value
        .chars()
        .all(|ch| ch.is_ascii_digit() || ch == '-' || ch == '.')
        && value.chars().any(|ch| ch.is_ascii_digit())
    {
        if value.contains('.') {
            if let Ok(number) = value.parse::<f64>() {
                return json!(number);
            }
        } else if let Ok(number) = value.parse::<i64>() {
            return json!(number);
        }
    }
    Value::from(value)
}

fn unquote(value: &str) -> String {
    if value.len() >= 2 {
        let bytes = value.as_bytes();
        if bytes[0] == bytes[value.len() - 1] && (bytes[0] == b'\'' || bytes[0] == b'"') {
            return value[1..value.len() - 1].to_string();
        }
    }
    value.to_string()
}

fn string_list(value: Option<&Value>) -> Vec<String> {
    match value {
        Some(Value::Array(items)) => items
            .iter()
            .filter_map(Value::as_str)
            .map(str::to_string)
            .collect(),
        _ => Vec::new(),
    }
}

pub fn list_skills_rpc(params: &Value) -> Result<Value, (i64, String, &'static str)> {
    let Some(roots_raw) = params.get("roots").and_then(Value::as_array) else {
        return Err((
            -32602,
            "skills.list: `roots` (array of paths) is required".into(),
            "invalid_params",
        ));
    };
    let roots: Vec<PathBuf> = roots_raw
        .iter()
        .filter_map(Value::as_str)
        .map(PathBuf::from)
        .collect();
    let conditions: Vec<String> = params
        .get("conditions")
        .and_then(Value::as_array)
        .map(|items| {
            items
                .iter()
                .filter_map(Value::as_str)
                .map(str::to_string)
                .collect()
        })
        .unwrap_or_default();
    let exclude: Vec<String> = params
        .get("exclude")
        .and_then(Value::as_array)
        .map(|items| {
            items
                .iter()
                .filter_map(Value::as_str)
                .map(str::to_string)
                .collect()
        })
        .unwrap_or_default();
    let ignore_conditions = params
        .get("ignoreConditions")
        .and_then(Value::as_bool)
        .unwrap_or(false);
    let provider = FilesystemSkillProvider::new(&roots);
    let selected = select_skills(provider.list(), &conditions, &exclude, ignore_conditions);
    Ok(json!({
        "skills": selected.iter().map(SkillDefinition::to_rpc).collect::<Vec<_>>(),
    }))
}
